include/stdx/span.hpp

```diff
@@ -0,0 +1,256 @@
+              std::enable_if_t<dependent_extent<U> != dynamic_extent and
+                                   N == dynamic_extent,
+                               int> = 0>
+    explicit constexpr span(span<U, N> const &s) noexcept : ptr{s.data()} {}
+
+    template <class U, std::size_t N,
+              std::enable_if_t<dependent_extent<U> == dynamic_extent or
+                                   N != dynamic_extent,
+                               int> = 0>
+    // NOLINTNEXTLINE(google-explicit-constructor)
+    constexpr span(span<U, N> const &s) noexcept
+        : base_t{s.data(), s.size()}, ptr{s.data()} {}
+
+    [[nodiscard]] constexpr auto data() const noexcept -> pointer {
+        return ptr;
+    }
+
+    [[nodiscard]] constexpr auto begin() const noexcept -> iterator {
+        return ptr;
+    }
+    [[nodiscard]] constexpr auto cbegin() const noexcept -> const_iterator {
+        return ptr;
+    }
+    [[nodiscard]] constexpr auto rbegin() const noexcept -> reverse_iterator {
+        return std::reverse_iterator{end()};
+    }
+    [[nodiscard]] constexpr auto crbegin() const noexcept
+        -> const_reverse_iterator {
+        return std::reverse_iterator{cend()};
+    }
+
+    [[nodiscard]] constexpr auto end() const noexcept -> iterator {
+        return ptr + this->size();
+    }
+    [[nodiscard]] constexpr auto cend() const noexcept -> const_iterator {
+        return ptr + this->size();
+    }
+    [[nodiscard]] constexpr auto rend() const noexcept -> reverse_iterator {
+        return std::reverse_iterator{begin()};
+    }
+    [[nodiscard]] constexpr auto crend() const noexcept
+        -> const_reverse_iterator {
+        return std::reverse_iterator{cbegin()};
+    }
+
+    [[nodiscard]] constexpr auto front() const -> reference { return *begin(); }
+    [[nodiscard]] constexpr auto back() const -> reference {
+        return *std::prev(end());
+    }
+
+    [[nodiscard]] constexpr auto operator[](size_type idx) const -> reference {
+        return data()[idx];
+    }
+
+    template <std::size_t Count>
+    [[nodiscard]] constexpr auto first() const -> span<element_type, Count> {
+        return span<element_type, Count>{ptr, Count};
+    }
+    [[nodiscard]] constexpr auto first(size_type count) const
+        -> span<element_type, dynamic_extent> {
+        return {ptr, count};
+    }
+
+    template <std::size_t Count>
+    [[nodiscard]] constexpr auto last() const -> span<element_type, Count> {
+        return span<element_type, Count>{ptr + this->size() - Count, Count};
+    }
+    [[nodiscard]] constexpr auto last(size_type count) const
+        -> span<element_type, dynamic_extent> {
+        return {ptr + this->size() - count, count};
+    }
+
+    template <std::size_t Offset, std::size_t Count = dynamic_extent>
+    [[nodiscard]] constexpr auto subspan() const {
+        if constexpr (Count != dynamic_extent) {
+            return span<element_type, Count>{ptr + Offset, Count};
```

Comment:
**suggestion**: ensure that `Count` and `Offset` form a valid pair when the `Extent` is not dynamic.